src/include/sof/audio/codec_adapter/codec/generic.h

```diff
@@ -0,0 +1,189 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ *
+ * \file generic.h
+ * \brief Generic Codec API header file
+ * \author Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ */
+
+#ifndef __SOF_AUDIO_CODEC_GENERIC__
+#define __SOF_AUDIO_CODEC_GENERIC__
+
+#include <sof/audio/component.h>
+
+#define comp_get_codec(d) (&(((struct comp_data *)((d)->priv_data))->codec))
+#define CODEC_GET_INTERFACE_ID(id) ((id) >> 0x8)
+#define CODEC_GET_API_ID(id) ((id) & 0xFF)
+#define API_CALL(cd, cmd, sub_cmd, value, ret) \
+	do { \
+		ret = (cd)->api((cd)->self, \
+				(cmd), \
+				(sub_cmd), \
+				(value)); \
+	} while (0)
+
+/*****************************************************************************/
+/* Codec generic data types						     */
+/*****************************************************************************/
+/**
+ * \struct codec_interface
+ * \brief Codec specific interfaces
+ */
+struct codec_interface {
+	/**
+	 * The unique ID for a codec, used for initialization as well as
+	 * parameters loading.
+	 */
+	uint32_t id;
+	/**
+	 * Codec specific initialization procedure, called as part of
+	 * codec_adapter component creation in .new()
+	 */
+	int (*init)(struct comp_dev *dev);
+	/**
+	 * Codec specific prepare procedure, called as part of codec_adapter
+	 * component
+	 * preparation in .prepare()
```

Comment:
Can be merged in above line. Btw, can we add what calls are optional and mandatory.

Reply:
Ok so, the waring in `COMP_CMD_SET_DATA` has been removed as this functionality has been implemented. 
Do you have some concerns with `comp_cl_info(&comp_codec_adapter, "codec_adapter_free(): component memory freed");` not sure what you meant by `ack` in this regards? The comment lin `generic.h` has been merged. As for what methods are mandatory - all of them are.